File: js/immutability.js

var Student = {
  courses: [],
  enroll: function(course) {
    course.startedAt = new Date()
    this.courses.push(course)
  }
}

function saveStudent(student) {
  console.log('Saving student... %j', student);
}

function saveCourseStats(course) {
  console.log('                                                    |')
  console.log('                                                    |')
  console.log('                                                    |')
  console.log('                                                    V')
  console.log('Updating course... %j', course);
  console.log('Updating stats...');
}

function getStudentById(sid) {
  var student = Object.create(Student)
  student.id = sid
  student.name = 'Marie Curie'
  return student
}

function getCourseByCode(code) {
  var courses = [
    {
      name: 'Math',
      code: 'ma-101'
    },
    {
      name: 'Art',
      code: 'ar-101'
    },
    {
      name: 'History',
      code: 'hi-101'
    }
  ]
  return courses.filter((c) => c.code == code)[0]
}

function enrollInCourse(studentId, courseCode) {
  var course = getCourseByCode(courseCode)
  var st = getStudentById(studentId)
  st.enroll(course)
  saveStudent(st)
  saveCourseStats(course)
}

enrollInCourse('1029384756', 'ma-101')

File: clj/src/immutability.clj
(ns immutability
  (:require [clojure.pprint :refer [pprint]]))

(defn enroll
  [student course]
  (->> (assoc course :started-at (java.util.Date.))
       (update student :courses conj)))

(defn enroll
  [student course]
  (let [enrolled-at (new java.util.Date)
        c (assoc course :started-at enrolled-at)
        courses (conj (:courses student) c)]
    (assoc student :courses courses)))

(defn student
  [sid name]
  {:courses []
   :id sid
   :name name})

(defn save-student [student]
  (println "Saving student...")
  (pprint student))

(defn save-course-stats
  [course]
  (println "Updating course " course)
  (println "Updating stats.."))

(defn get-student-by-id
  [sid]
  (student sid "Marie Curie"))

(defn get-course-by-code
  [code]
  (let [courses [{:name "Math"
                  :code "ma-101"}
                 {:name "Art"
                  :code "ar-101"}
                 {:name "History"
                  :code "hi-101"}]]
    (first (filter #(= (:code %) code) courses))))

(defn enroll-in-course
  [student-id course-code]
  (let [course (get-course-by-code course-code)
        st (-> (get-student-by-id student-id)
               (enroll course)
               (save-student))]
    (save-course-stats course)))

(defn -main
  [& args]
  (enroll-in-course "1029384756" "ma-101"))
